### PR TITLE
Task 1.1 Prefetching in Simple FSDP

### PR DESCRIPTION
**Pull request by ruisizhang123**
Add code for Task 1.1 Prefetching in Simple FSDP

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 from torch.utils._sympy.symbol import free_symbol_is_type, SymT
 from torch.utils._triton import has_triton
 
-from . import comms, config, dependencies, ir, metrics
+from . import comms, config, dependencies, ir, metrics, spmd_prefetch, spmd_bucketing
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py]: IMO there's no need to have two separate files, we can put them into one file and maybe call it `simple_fsdp`

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -1477,7 +1477,15 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
         self.create_foreach_nodes()
         self.topological_sort_schedule()
         self.logged_slow_fusion: Set[Tuple[str, str]] = set()
+        self.nodes = spmd_bucketing.bucketing_per_blcok(self.nodes)
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py]: let's remove this as it is not relevant to this PR -- let's address one problem at a time.

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -1477,7 +1477,15 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
...
         self.fuse_nodes()
+        # do prefetching reordering
+        if self.post_grad_graph_id == 0:
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py:1482]: Does id=0 means forward graph and id=1 means backward graph?
If so please comment out in the code to make it more readable.

File: torch/_inductor/spmd_prefetch.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
+
+
+    
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_order: Optional[str] = "before",
```
Comment (tianyu-l) [inline torch/_inductor/spmd_prefetch.py]: Since this is a binary option, it might be better to use a boolean variable to denote.
E.g. call it `all_gather_before_last_wait = True`

File: torch/_inductor/spmd_prefetch.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            wait_list.append(node)
+        else:
+            raise ValueError("node type not supported")
+
+    if len(wait_list) > 0:
+        reorder_list.extend(wait_list)
+    return reorder_list
+
+
+def get_node_type(node, prev_nodes=None) -> int:
```
Comment (tianyu-l) [inline torch/_inductor/spmd_prefetch.py]: We should use Enum (with meaningful names) instead of int for node categorization to make code more readable.

File: torch/_inductor/spmd_prefetch.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    '''
+    Reorder Reduce Scatter and Wait in the backward pass
+    reorder wait_i_rs before reduce_scatter_i+1 
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    wait_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+    
+    for node in snodes:
+        # ruisi: to get the convert_element_type after rs, can be removed after fixing the mpt bug.
```
Comment (tianyu-l) [inline torch/_inductor/spmd_prefetch.py]: for these type of temporary comments, a common practice is to use `# TODO: ` or `# TODO(ruisizhang123)` when you would like to follow up later.

File: torch/_inductor/spmd_prefetch.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
+
+    inverse_users, node_users = comms.compute_node_users(snodes)
+
+    for node in reversed(snodes):
+        node_type_list.append(get_node_type(node))
+
+    for idx, node in enumerate(reversed(snodes)):
```
Comment (tianyu-l) [inline torch/_inductor/spmd_prefetch.py]: The two reordering algorithms look non-straightforward, especially with the integer node types.
After changing node types to use Enum, would you please add comments to code to explain how you achieve the desired order? Good code readability makes it easy to understand and debug.

File: torch/utils/data/dataloader.py [hunk 8]
```diff
@@ -44,6 +44,8 @@
 ]
 
 
+T_co = TypeVar('T_co', covariant=True)
+T = TypeVar('T')
```
Comment (tianyu-l) [inline torch/utils/data/dataloader.py]: This is being fixed https://github.com/pytorch/data/pull/1282
After that PR landed, you can reinstall the latest torchdata nightly to avoid manually changing it here.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Great job on your first PR!
I left some comments, please see if they make sense.

File: torch/_inductor/scheduler.py [hunk 9]
```diff
@@ -1477,7 +1477,17 @@ def __init__(self, nodes: List[ir.Operation]) -> None:
         self.create_foreach_nodes()
         self.topological_sort_schedule()
         self.logged_slow_fusion: Set[Tuple[str, str]] = set()
+        #self.nodes = simple_fsdp.bucketing_per_blcok(self.nodes)
```
Comment (tianyu-l) [inline torch/_inductor/scheduler.py]: let's remove this line to avoid leakage of future changes

File: torch/_inductor/simple_fsdp.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        if "wait_tensor" in node_origins_name:
+            node_type = NodeType.WAIT
+        if "reduce_scatter_tensor" in node_origins_name:
+            node_type = NodeType.REDUCE_SCATTER
+        if "convert_element_type" in node_origins_name and prev_nodes == [NodeType.REDUCE_SCATTER, NodeType.WAIT]:
+            node_type = NodeType.COVERT_ELEMENT
+    return node_type
+
+
+def bucketing_per_blcok(
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: ditto: pls remove it from this PR

File: torch/_inductor/simple_fsdp.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    if not isinstance(node, scheduler.FusedSchedulerNode):
+        node_origins_name = " ".join([str(i.name) for i in list(node.node.origins)])
+        if "all_gather_into_tensor" in node_origins_name:
+            node_type = NodeType.ALL_GATHER
+        if "wait_tensor" in node_origins_name:
+            node_type = NodeType.WAIT
+        if "reduce_scatter_tensor" in node_origins_name:
+            node_type = NodeType.REDUCE_SCATTER
+        if "convert_element_type" in node_origins_name and prev_nodes == [NodeType.REDUCE_SCATTER, NodeType.WAIT]:
+            node_type = NodeType.COVERT_ELEMENT
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: typo: `COVERT_ELEMENT` should be `CONVERT_ELEMENT_TYPE`

File: torch/_inductor/simple_fsdp.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            if not all_gather_before_last_wait and node_type_list[idx+1]==NodeType.ALL_GATHER and len(all_gather_list) > 0:
+                # move i-th all gather node and its dependencies after (i-1)-th wait node (bc this is a reverse list)
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
+            # add wait node
+            reorder_list.append(node)
+            if all_gather_before_last_wait and node_type_list[idx+1]==NodeType.ALL_GATHER and len(all_gather_list) > 0: 
+                # move i-th all gather node and its dependencies before (i-1)-th wait node (bc this is a reverse list)
+                reorder_list.extend(all_gather_list)
+                all_gather_list = []
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: I think it would be better if we restructure it:
1. first tell if `node_type_list[idx+1]==NodeType.ALL_GATHER` -- if not we just do append
2. otherwise we test if `all_gather_before_last_wait`

Do we need to ask `len(all_gather_list) > 0`?
If we expect it to always be nonzero, we should do an assertion here.

File: torch/_inductor/simple_fsdp.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    node_type = NodeType.COMP
+    if not isinstance(node, scheduler.FusedSchedulerNode):
+        node_origins_name = " ".join([str(i.name) for i in list(node.node.origins)])
+        if "all_gather_into_tensor" in node_origins_name:
+            node_type = NodeType.ALL_GATHER
+        if "wait_tensor" in node_origins_name:
+            node_type = NodeType.WAIT
+        if "reduce_scatter_tensor" in node_origins_name:
+            node_type = NodeType.REDUCE_SCATTER
+        if "convert_element_type" in node_origins_name and prev_nodes == [NodeType.REDUCE_SCATTER, NodeType.WAIT]:
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: we should move the TODO to right before this line

File: torch/_inductor/simple_fsdp.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_type_list: List[int] = []
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: would it be more readable if we make `node_type_list` into a dictionary?

File: torch/_inductor/simple_fsdp.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+def reorder_all_gather(
+    snodes: List["scheduler.BaseSchedulerNode"],
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    '''
+    reorder_list: List["scheduler.BaseSchedulerNode"] = []
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: consider renaming it to something like `results` to differentiate it from other intermediate lists.

File: torch/_inductor/simple_fsdp.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        reorder_list.extend(wait_list)
+    return reorder_list
+
+
+def get_node_type(node, prev_nodes=None) -> int:
+    # node_type: {0: all gather; 1: wait_tensor; 2: computation; 3: reduce scatter; 4: convert_element_type after rs}
+    # TODO(ruisizhang123): we add [4: convert_element_type] bc of some bugs in mpt conversion. after tianyu fixed this, we can remove it
+    node_type = NodeType.COMP
+    if not isinstance(node, scheduler.FusedSchedulerNode):
+        node_origins_name = " ".join([str(i.name) for i in list(node.node.origins)])
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: As we discussed offline, can we use more direct ways to tell what the node type is, e.g. using a node's kernel type?

File: torch/_inductor/simple_fsdp.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+                all_gather_list = []
+        else:
+            raise ValueError("node type not supported")
+    if len(all_gather_list) > 0:
+        reorder_list.extend(all_gather_list)
+    reorder_list.reverse()
+    return reorder_list
+
+
+
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py:73]: Let's use only 2 empty lines between functions. In fact, you can install `lintrunner` to help you do linting.

**Review by tianyu-l — COMMENTED** [review]
Nice improvements to the first version! The code becomes much more readable. I left some more comments. Let's address them before landing.

File: torch/_inductor/simple_fsdp.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    all_gather_before_last_wait: Optional[bool] = True,
+) -> List["scheduler.BaseSchedulerNode"]:
+    '''
+    Reorder All Gather and Wait in the forward/backward pass;
+    1. all_gather_before_last_wait set to True: all_gather_i is reordered before wait_i-1
+    2. all_gather_before_last_wait set to False: all_gather_i is reordered after wait_i-1
+    '''
+    result_list: List["scheduler.BaseSchedulerNode"] = []
+    all_gather_list: List["scheduler.BaseSchedulerNode"] = []
+    node_name_to_type: OrderedDict[str: NODE_TYPE] = {}
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: any reason we need an OrderedDict instead of just a dict here?

Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: I wonder can the keys just be nodes instead of their names?

File: torch/_inductor/simple_fsdp.py [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+    Union,
+)
+from enum import IntEnum
+from . import config, ir, comms, scheduler
+from collections import OrderedDict
+
+import torch
+
+from .dependencies import WeakDep
+from .utils import is_collective, is_wait, tuple_sorted
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: let's use `lintrunner` to automatically group the imports better

File: torch/_inductor/simple_fsdp.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+
+class NODE_TYPE(IntEnum):
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: It seems the standard way should be `NodeType` according to https://docs.python.org/3/howto/enum.html

File: torch/_inductor/simple_fsdp.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+    result_list: List["scheduler.BaseSchedulerNode"] = []
+    wait_list: List["scheduler.BaseSchedulerNode"] = []
+    node_name_to_type: OrderedDict[str: NODE_TYPE] = {}
+    
+    prev_node_type = []
+    for node in snodes:
+        # TODO(ruisizhang123): to get the convert_element_type after rs, can be removed after fixing the mpt bug.
+        cur_node_type = get_node_type(node, prev_node_type[-2:])
+        node_name_to_type[node.get_name()] = cur_node_type
+        prev_node_type.append(cur_node_type)
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: keeping a list is overkill, but it might be fine since we'll remove it after fixing the todo issue anyway.

**Review by tianyu-l — APPROVED** [review]
LGTM! Feel free to merge.

File: torch/_inductor/simple_fsdp.py [hunk 22]
```diff
@@ -83,29 +82,25 @@ def reorder_reduce_scatter(
     result_list: List[scheduler.BaseSchedulerNode] = []
     wait_list: List[scheduler.BaseSchedulerNode] = []
     node_to_type: Dict[scheduler.BaseSchedulerNode, int] = {}
-
+    inverse_users, node_users = comms.compute_node_users(snodes)
     prev_node_type: List[int] = []
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: can be removed

File: torch/_inductor/simple_fsdp.py [hunk 23]
```diff
@@ -83,29 +82,25 @@ def reorder_reduce_scatter(
...
-        node_to_type[node] = cur_node_type
-        prev_node_type.append(cur_node_type)
+        node_to_type[node] = get_node_type(node)
 
     for idx, node in enumerate(snodes):
         node_type = node_to_type[node]
         if node_type in [NodeType.ALL_GATHER, NodeType.COMPUTE]:
             # we do not reorder all gather and compute node
-            result_list.append(node)
+            if node not in result_list and node not in wait_list:
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: it is true that we only need to test `if node not in wait_list:`

File: torch/_inductor/simple_fsdp.py [hunk 24]
```diff
@@ -122,7 +117,7 @@ def reorder_reduce_scatter(
 
 
 def get_node_type(node, prev_nodes=None) -> int:
-    # node_type: {0: all gather; 1: wait_tensor; 2: computation; 3: reduce scatter; 4: convert_element_type after rs}
+    # node_type: {0: all gather; 1: wait_tensor; 2: computation; 3: reduce scatter;}
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py]: Let's remove this comment; otherwise we need to maintain it if the NodeType changes.

File: torch/_inductor/simple_fsdp.py [hunk 25]
```diff
@@ -122,7 +117,7 @@ def reorder_reduce_scatter(
...
     node_type = NodeType.COMPUTE
     if not isinstance(node, scheduler.FusedSchedulerNode):
         if (
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp.py:124]: Let's use `if-else` to refactor the code in this function. Otherwise every `if` needs to be run.

File: torch/_inductor/simple_fsdp.py (hunk 22 above)
Comment (ruisizhang123) [reply torch/_inductor/simple_fsdp.py:86]: we introduce this line here bc we need to get wait's dependency node here. (I'll fix other things): https://github.com/ruisizhang123/pytorch_intern24/blob/8b7b39626a6b7faf8abe5c337c65ae982ea8c354/torch/_inductor/simple_fsdp.py#L100
